src/index.js

```diff
@@ -1,174 +1,226 @@
+   * 
+   * @param {String} [id] - spark's ID
+   * @return {Promise.<String[]>} containing room names
    * @api public
    */
-
-  function get(id, fn = noop) {
-    let rms = keys(id ? sids[id] || {} : rooms)
-    setImmediate(fn, null, rms)
+  function get(id) {
+    if (id) {
+      return Promise.resolve(Array.from(sids.get(id)));
+    }
+    return Promise.resolve(Array.from(rooms.keys()));
   }
 
   /**
    * Add a socket to a room.
    *
-   * @param {String} id
-   * @param {String} room
-   * @param {Function} [fn]
+   * @param {String} id - spark's ID
+   * @param {String} room - the room to join
+   * @return {Promise} with no data
    * @api public
    */
-
-  function add(id, room, fn = noop) {
-    sids[id] = sids[id] || []
-    sids[id][room] = true
-    rooms[room] = rooms[room] || {}
-    rooms[room][id] = true
-    wild.add(room)
-    setImmediate(fn, null, null)
+  function add(id, room) {
+    const socketRooms = sids.get(id);
+    const socketsInRoom = rooms.get(room);
+    // if socket has a rooms set just add the room
+    if (socketRooms) {
+      socketRooms.add(room);
+    } else {
+      // socket has no rooms set, create and add the room
+      sids.set(id, new Set([room]))
+    }
+    // if room has a set of sockets, just add the new socket id
+    if (socketsInRoom) {
+      socketsInRoom.add(id);
+    } else {
+      // room has no set of sockets, create and add it
+      rooms.set(room, new Set([id]));
+    }
+    // add to wild rooms
+    // TODO visit wild logic to work Map and Set
+    // wild.add(room);
+    return Promise.resolve();
   }
 
   /**
    * Remove a socket from a room or from all rooms if a room is not passed.
-   *
-   * @param {String} id
-   * @param {String} room
-   * @param {Function} [fn]
+   * 
+   * @param {String} id - spark's ID
+   * @param {String} [room] - the room to leave
+   * @return {Promise} no data
    * @api public
    */
-
-  function del(id, room, fn = noop) {
-    let prune = pruner(id)
+  function del(id, room) {
+    // if (room) {
+    //   // TODO update logic for wildcard
+    //   if (wildcard && ~room.indexOf('*')) {
+    //     let rms = keys(id ? sids[id] || {} : rooms)
+    //     wild.find(room, rms, prune)
+    //   } else if (sids[id] && sids[id][room]) {
+    //     prune(room)
+    //   }
+    // } else {
+    //   for (room in sids[id]) prune(room)
+    // }
+    // setImmediate(fn, null, null)
+    
     if (room) {
-      if (wildcard && ~room.indexOf('*')) {
-        let rms = keys(id ? sids[id] || {} : rooms)
-        wild.find(room, rms, prune)
-      } else if (sids[id] && sids[id][room]) {
-        prune(room)
-      }
-    } else {
-      for (room in sids[id]) prune(room)
+      return prune(id, room);
     }
-    setImmediate(fn, null, null)
+    const promises = [];
+    rooms.forEach((value, key) => {
+      promises.push(prune(id, key));
+    });
+    return Promise.all(promises);
```

Comment:
We should add a `.then(noop)` to ensure that the promise will resolve to `undefined` and not an empty array.